src/test/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPluginTest.java

```diff
@@ -0,0 +1,191 @@
+
+    spyPlugin.openInitialConnection(mockConnectionUrl);
+
+    verify(mockSecretCache).get(SECRETS_MANAGER_SECRET_ID);
+    verify(spyPlugin).getCurrentCredentials();
+    verify(mockSecretCache).put(eq(SECRETS_MANAGER_SECRET_ID), any(String.class));
+    verify(mockPlugin, times(2)).openInitialConnection(captor.capture());
+    final List<ConnectionUrl> connectionUrls = captor.getAllValues();
+    // Verify the openInitialConnection method was called using different ConnectionUrl arguments
+    assertNotEquals(connectionUrls.get(0).toString(), connectionUrls.get(1).toString());
+  }
+
+  @Test
+  public void testFailedToReadSecrets() throws SQLException {
+    // Fail initial connection attempt so secrets will be retrieved
+    final SQLException failedFirstConnectionAccessException  = new SQLException(SQL_ERROR, AWSSecretsManagerPlugin.SQLSTATE_ACCESS_ERROR);
+    doThrow(failedFirstConnectionAccessException).doNothing().when(mockPlugin).openInitialConnection(any(ConnectionUrl.class));
+    when(mockSecretCache.get(any(String.class))).thenReturn("");
+    doReturn("{username: invalid, password: invalid}").when(spyPlugin).getCurrentCredentials();
+
+    final SQLException failedReadSecretsException = assertThrows(SQLException.class, () -> spyPlugin.openInitialConnection(mockConnectionUrl));
+
+    assertEquals(failedReadSecretsException.getMessage(), AWSSecretsManagerPlugin.ERROR_GET_SECRETS_FAILED);
+    verify(mockSecretCache).get(SECRETS_MANAGER_SECRET_ID);
+    verify(spyPlugin, times(2)).updateConnectionProperties(any(Properties.class), any(String.class));
+    verify(spyPlugin).getCurrentCredentials();
+    verify(mockPlugin).openInitialConnection(any());
+  }
+
+  @Test
+  public void testFailedToGetSecrets() throws SQLException {
+    // Fail initial connection attempt so secrets will be retrieved
+    final SQLException failedFirstConnectionAccessException  = new SQLException(SQL_ERROR, AWSSecretsManagerPlugin.SQLSTATE_ACCESS_ERROR);
+    doThrow(failedFirstConnectionAccessException).doNothing().when(mockPlugin).openInitialConnection(any(ConnectionUrl.class));
+    final SQLException getSecretsFailedException = new SQLException(AWSSecretsManagerPlugin.ERROR_GET_SECRETS_FAILED);
+    when(mockSecretCache.get(any(String.class))).thenReturn("");
+    doThrow(getSecretsFailedException).when(spyPlugin).getCurrentCredentials();
+
+    final SQLException connectionFailedException = assertThrows(SQLException.class, () -> spyPlugin.openInitialConnection(mockConnectionUrl));
+
+    assertEquals(connectionFailedException.getMessage(), AWSSecretsManagerPlugin.ERROR_GET_SECRETS_FAILED);
+    verify(mockSecretCache).get(SECRETS_MANAGER_SECRET_ID);
+    verify(spyPlugin).updateConnectionProperties(any(Properties.class), any(String.class));
+    verify(spyPlugin).getCurrentCredentials();
+    verify(mockPlugin).openInitialConnection(any());
+  }
+
+  private void initializePlugin() throws SQLException {
```

Comment:
It's simple logic in this method. Extracting it to a function makes code less readable. And the function is used in just a single place so I don't see many points to have it.